# Day 3 - Week 3: tasks

### Importing

In [21]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LogisticRegression

from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix, classification_report
from sklearn.metrics import roc_auc_score, roc_curve
import matplotlib.pyplot as plt


###  Step 1: Train a LogisticRegression model on a provided classification dataset (e.g. churn).
Loaded the titanic dataset

In [22]:
df = pd.read_csv('../../week1/day4/train.csv')
df.drop(['PassengerId','Name','Ticket','Cabin'],axis = 1,inplace = True)
df =pd.get_dummies(df)
df['Age']=df['Age'].fillna(df['Age'].mean())


target = 'Survived'
X = df.drop(target,axis=1)
y = df[target]

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42)

model = LogisticRegression(max_iter=1000)
model.fit(X_train, y_train)
pred = model.predict(X_test)
prob = model.predict_proba(X_test)[:,1]   



### Step 2: Generate predictions and produce the confusion matrix.

In [23]:
cm = confusion_matrix(y_test, pred)

print(f'TN = {cm[0][0]}  .  FP = {cm[0][1]}')
print(f'TP = {cm[1][0]}  .  FN = {cm[1][1]}')



TN = 90  .  FP = 15
TP = 19  .  FN = 55


### Step 3: Compute precision, recall, and F1 with classification_report, and interpret each.

In [24]:
print(classification_report(y_test, pred))

              precision    recall  f1-score   support

           0       0.83      0.86      0.84       105
           1       0.79      0.74      0.76        74

    accuracy                           0.81       179
   macro avg       0.81      0.80      0.80       179
weighted avg       0.81      0.81      0.81       179



The model achieved `81%` accuracy. It performed a little bit better on **class 0** than on **class 1**. For class 1, the precision was `0.79`, recall was `0.74`, and the F1-score was `0.76`, it shows good overall performance but with some missed positive cases.

### Step 4: Decide whether precision or recall matters more for this specific problem and justify it.

Recall is more important for the Titanic dataset problem because we want the model to find as many survivors as possible.False Negative happens when the model predicts that a passenger didn't survive, but in reality they did survive. Missing these cases means the model failed to identify some of the actual survivors. Therefore, focusing on improving recall helps the model detect more true survivors, even if it leads to some incorrect predictions of survival.

### Step 5: Compute the AUC-ROC and document what it says about the model.

In [26]:
auc = roc_auc_score(y_test, prob)

print("AUC:", auc)

AUC: 0.8823680823680825


### AUC-ROC Evaluation

The model achieved an AUC-ROC score of **0.88**. This indicates that the model has a good ability to distinguish between passengers who survived and those who didn't. The AUC value ranges from 0.5 (random guessing) to 1.0 (perfect classification), so a score of 0.88 shows that the model performs much better than random guessing and can effectively separate the two classes.